src/BatchRestore/Program.cs

```diff
@@ -0,0 +1,147 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Fx.CommandLine;
+
+namespace BatchRestore
+{
+    /// <summary>
+    /// This simple command works around a DNX issue where it will not recur under a directory which
+    /// contains a project.json when restoring packages. 
+    /// (see https://github.com/aspnet/dnx/commit/0dda8bf86863364cc20421f1af7494f1b2a3256f for more context).
+    /// 
+    /// This program takes a list of directories to restore (and optionally a list of sources and a pacakges
+    /// folder) and computes all the directories that will need to be restored so that all project.json's are
+    /// restored. We then call out to dotnet-restore to do the actual work of restoring. To minimize the number
+    /// of instances of dotnet-restore we need to launch we batch up folders to pass to each invocation, being
+    /// mindful of the command line limits of Windows.
+    /// 
+    /// This program needs to be placed next to `dotnet-restore` and can then be launched with 
+    /// `dotnet batch-restore`.  Note we use CoreFX command line argument styles (i.e. -foo instead of
+    /// --foo).
+    /// </summary>
+    internal class Program
+    {
+        private static List<string> s_directoriesToRestore = new List<string>();
+        private const int CommandLineLengthLimit = 7500; /* Well under the 8191 limit of Windows */
+
+        private static int Main(string[] args)
+        {
+            ParseCommandLine(args);
+
+            foreach (string directory in s_rootDirectories)
+            {
+                VisitDirectory(directory, true);
+            }
+
+            string argumentPrefix = GetDotnetRestoreCommandPrefix();
+
+            // Loop over all the directories to restore, invoking dotnet-restore as few times as possible
+            // by batching things together.
+            StringBuilder argumentBuilder = new StringBuilder(argumentPrefix, CommandLineLengthLimit);
+            foreach(string directoryToRestore in s_directoriesToRestore)
+            {
+                argumentBuilder.Append(" \"");
+                argumentBuilder.Append(directoryToRestore);
+                argumentBuilder.Append('"');
+
+                if (argumentBuilder.Length > CommandLineLengthLimit)
+                {
+                    int retValue = RunDotnetRestore(argumentBuilder);
+
+                    if (retValue != 0)
+                    {
+                        return retValue;
+                    }
+
+                    argumentBuilder.Clear();
+                    argumentBuilder.Append(argumentPrefix);
+                }
+            }
+
+            if (argumentBuilder.Length > argumentPrefix.Length)
+            {
+                return RunDotnetRestore(argumentBuilder);
+            }
+
+            return 0;
+        }
+
+        private static int RunDotnetRestore(StringBuilder argumentBuilder)
```

Comment:
Nit: since no arguments are being added in this method, seems like this could just be ```string arguments```.